lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -2135,6 +2182,236 @@ LogicalResult WhileOp::canonicalize(WhileOp whileOp,
+  // The done port of these instances is the last result, so the ssa value
+  // of the last result is taken.
+  if (isa<RegisterOp, MemoryOp, DivSPipeLibOp, DivUPipeLibOp, MultPipeLibOp,
+          RemSPipeLibOp, RemUPipeLibOp>(operation)) {
+    size_t doneIdx = operation->getResults().size() - 1;
+    ret = operation->getResult(doneIdx);
+  } else if (isa<InstanceOp>(operation)) {
+    //  Get the go port of the instance through the "done" attribute
+    InstanceOp instanceOp = cast<InstanceOp>(operation);
+    auto portInfo = instanceOp.getReferencedComponent().getPortInfo();
+    for (size_t i = 0; i != portInfo.size(); i++) {
+      if (portInfo[i].hasAttribute("done"))
+        ret = operation->getResult(i);
+    }
+  } else if (isa<calyx::PrimitiveOp>(operation)) {
+    // Get the go part of the primitive through the "calyx.done" attribute
+    PrimitiveOp primOp = cast<PrimitiveOp>(operation);
+    auto moduleExternOp = primOp.getReferencedPrimitive();
+    auto resAttrs = moduleExternOp.getResAttrsAttr();
+    for (size_t i = 0; i != resAttrs.size(); i++)
+      if (DictionaryAttr dictAttr = dyn_cast<DictionaryAttr>(resAttrs[i]))
+        if (!dictAttr.empty())
+          if (dictAttr.begin()->getName().getValue() == "calyx.done")
+            ret = primOp.getResult(i);
+  }
+  return ret;
+}
+
+LogicalResult InvokeOp::verify() {
+  ComponentOp componentOp = (*this)->getParentOfType<ComponentOp>();
+  StringRef callee = getCallee();
+  Operation *operation = componentOp.lookupSymbol(callee);
+  // The referenced symbol does not exist.
+  if (!operation)
+    return emitOpError() << "with instance '" << callee
+                         << "', which does not exist.";
+  // The argument list of invoke is empty.
+  if (getInputs().empty())
+    return emitOpError() << "the input for '" << getOperationName()
+                         << "' is empty.";
+  size_t goPortNum = 0, donePortNum = 0;
+  // They both have a go port and a done port, but the "go" port for
+  // registers and memrey should be "write_en" port.
+  if (isa<RegisterOp, DivSPipeLibOp, DivUPipeLibOp, MemoryOp, MultPipeLibOp,
+          RemSPipeLibOp, RemUPipeLibOp>(operation))
+    goPortNum = 1, donePortNum = 1;
+  else if (InstanceOp instanceOp = dyn_cast<InstanceOp>(operation)) {
+    auto portInfo = instanceOp.getReferencedComponent().getPortInfo();
+    for (PortInfo info : portInfo) {
+      if (info.hasAttribute("go"))
+        goPortNum++;
+      if (info.hasAttribute("done"))
+        donePortNum++;
+    }
+  } else if (PrimitiveOp primOp = dyn_cast<PrimitiveOp>(operation)) {
+    auto moduleExternOp = primOp.getReferencedPrimitive();
+    // Get the number of go ports and done ports by their attrubutes.
+    for (Attribute attr : moduleExternOp.getArgAttrsAttr())
+      if (DictionaryAttr dictAttr = dyn_cast<DictionaryAttr>(attr))
+        if (!dictAttr.empty())
+          if (dictAttr.begin()->getName().getValue() == "calyx.go")
+            goPortNum++;
```

Comment:
This `for (...) if (...) if (...)` is duplicated here and below for two different strings. Please write a helper function. I think this could also probably be done with some C++ standard library algorithm, e.g., `llvm::count_if`